### DISCUSSION TITLE
How to use schema.arg to expose autogenerate "where" input arg from Prisma/Nexus types?

### DISCUSSION THREAD

**Discussion by exactstaff** [prisma/prisma#2725] [Q&A] [answered]
Hi,

I was wondering if there's a recommended way to expose the WhereInput fields or further more all the input fields pertaining to a certain model generated by prisma/nexus.

For example, I'm trying to do the following to take advantage of the autogenerated input type "CategoryWhereInput".  The were argument of this resolver is optional, if it's not provided it should return all the Categories. When I try to do the below, the IDE complains that, for example, Types of property **id** are incompatible,  type "null" is not assignable to type 'string | StringFilter | undefined'. 

```typescript
import { schema } from 'nexus'
import { core } from 'nexus/components/schema'

export default (t: core.ObjectDefinitionBlock<'Query'>) => {
  t.crud.category()
  t.list.field('categories', {
    type: 'Category',
    args: {
      where: schema.arg({ type: 'CategoryWhereInput' }),
      take: schema.intArg(),
    },
    resolve: async (parent, args, ctx) => {
      return ctx.db.category.findMany({
        where: { ...args.where }, ///IDE complais, due to the possibility of null values from argument
      })
    },
  })
}

**Comment by ryands17** [comment]
Hey @exactstaff :wave: 
Is this `CategoryWhereInput` created anywhere? If so, could you show that as well.

I think Nexus doesn't auto-generate types without explicity passing any `ordering` or `filtering` argument, so you would have to construct that argument yourself.

**Reply by exactstaff** [reply]
Hi @ryands17 

I believe that this is generated thanks to the nexus plugin prisma package. This is how I've using them to take advantage of the autogenerate input types. 

As you can see in the screenshot, it exposes the correct fields to allow filtering, but for some reason is not compatible with the where argument that prisma is expecting. 

![image](https://user-images.githubusercontent.com/56694175/84523682-9649b300-ac8d-11ea-987c-6ce6a51f46ed.png)

![image](https://user-images.githubusercontent.com/56694175/84524835-70251280-ac8f-11ea-8b76-16ed37fa8a4d.png)

**Reply by exactstaff** [reply]
It seems like it's defined in the types generated by the prisma client
![image](https://user-images.githubusercontent.com/56694175/84526582-75379100-ac92-11ea-869c-660b7649ca67.png)

**Comment by exactstaff** [comment]
Just to add a little bit more information. I'm reproducing the issue with a more simple argument. I added an argument of type **schema.stringArg()** named "name".

```typescript
import { schema } from 'nexus'
import { core } from 'nexus/components/schema'

export default (t: core.ObjectDefinitionBlock<'Query'>) => {
  t.crud.category()
  t.list.field('categories', {
    type: 'Category',
    args: {
      take: schema.intArg(),
      name: schema.stringArg(),
    },
    resolve: async (parent, args, ctx) => {
      return ctx.db.category.findMany({
        where: { name: args.name },
        ...(args.take ? { take: args.take } : {}),
      })
    },
  })
}

```

![image](https://user-images.githubusercontent.com/56694175/84536464-c0f23680-aca2-11ea-9aac-eba657ad85de.png)

It seems like the posible types for the argument are **string | null | undefined** but the property "name" that prisma is expecting only accepts **string | StringFilter | undefined**, so the fact that the argument could be null is causing the error message from TypeScript. 

If I set the option **nullable:false** to the argument, the error goes away, but I don't want to do it this way for an optional field.

```typescript
import { schema } from 'nexus'
import { core } from 'nexus/components/schema'

export default (t: core.ObjectDefinitionBlock<'Query'>) => {
  t.crud.category()
  t.list.field('categories', {
    type: 'Category',
    args: {
      take: schema.intArg(),
      name: schema.stringArg({ nullable: false }),
    },
    resolve: async (parent, args, ctx) => {
      return ctx.db.category.findMany({
        where: { name: args.name },
        ...(args.take ? { take: args.take } : {}),
      })
    },
  })
}
```

How would you approach this issue? if the argument is not present on the query should its type be undefined to make it compatible with the types that prisma expecting?

**Reply by ryands17** [reply, marked as answer]
@exactstaff One way of doing this would be setting `nullable: false` and not passing the field from the frontend if null, so that could be one case.

The other way would be if you want to set `nullable: true` would be to do something like this:

```ts
import { schema } from 'nexus'
import { core } from 'nexus/components/schema'

export default (t: core.ObjectDefinitionBlock<'Query'>) => {
  t.crud.category()
  t.list.field('categories', {
    type: 'Category',
    args: {
      take: schema.intArg(),
      name: schema.stringArg({ nullable: false }),
    },
    resolve: async (parent, args, ctx) => {
      return ctx.db.category.findMany({
        where: args.name ? { name: args.name } : {},
        ...(args.take ? { take: args.take } : {}),
      })
    },
  })
}
```

I could think of these two possibilities. Let me know if this works for you.

**Comment by ludralph** [comment]
Hi @exactstaff,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
